From 1eb77c4fc1786f4d58eccbded53774bffb0a3da1 Mon Sep 17 00:00:00 2001
From: Jiangjin Wang <kaymw@aosc.io>
Date: Sat, 11 Nov 2023 01:07:57 -0800
Subject: [PATCH] overlay: enable compilation on non-x86 platforms

- Add rmb/wmb definition for AArch64, LoongArch, MIPS64 and RISC-V. Fallback to
  nop if they are not defined.
---
 overlay/gpu-perf.c  | 36 ++++++++++++++++++++++++++++++++++++
 overlay/meson.build |  2 +-
 2 files changed, 37 insertions(+), 1 deletion(-)

diff --git a/overlay/gpu-perf.c b/overlay/gpu-perf.c
index 7addd78dd..7fae22f4b 100644
--- a/overlay/gpu-perf.c
+++ b/overlay/gpu-perf.c
@@ -50,6 +50,42 @@
 #define wmb()           asm volatile("sfence" ::: "memory")
 #endif
 
+#if defined(__aarch64__)
+#define dsb(opt)        asm volatile("dsb " #opt : : : "memory")
+#define rmb()           dsb(ld)
+#define wmb()           dsb(st)
+#endif
+
+#if defined(__loongarch__)
+#define DBAR(hint)      asm volatile("dbar %0 " : : "I"(hint) : "memory")
+#define cr_r_           0b00101
+#define c_w_w           0b01010
+#define c_rsync()       DBAR(cr_r_)
+#define c_wsync()       DBAR(c_w_w)
+#define rmb()           c_rsync()
+#define wmb()           c_wsync()
+#endif
+
+#if defined(__mips64)
+#define rmb()           asm volatile("sync" ::)
+#define wmb()           asm volatile("sync" ::)
+#endif
+
+#if defined(__riscv)
+#define RISCV_FENCE(p, s) \
+        __asm__ __volatile__ ("fence " #p "," #s : : : "memory")
+#define rmb()           RISCV_FENCE(ir,ir)
+#define wmb()           RISCV_FENCE(ow,ow)
+#endif
+
+#if ! defined(rmb)
+#define rmb()
+#endif
+
+#if ! defined(wmb)
+#define wmb()
+#endif
+
 #define N_PAGES 32
 
 struct sample_event {
diff --git a/overlay/meson.build b/overlay/meson.build
index 0a9907620..e0a3f2f82 100644
--- a/overlay/meson.build
+++ b/overlay/meson.build
@@ -83,7 +83,7 @@ else
 	message('WARNING: leg command not found, disabling overlay; try : apt-get install peg')
 endif
 
-if not build_overlay.disabled() and ['x86', 'x86_64'].contains(host_machine.cpu_family()) and libdrm_intel.found() and leg.found() and xrandr.found() and cairo.found() and (with_xlib_backend or with_xv_backend)
+if build_overlay.enabled() and libdrm_intel.found() and leg.found() and xrandr.found() and cairo.found() and (with_xlib_backend or with_xv_backend)
 	executable('intel-gpu-overlay', gpu_overlay_src,
 			include_directories : inc,
 			c_args : gpu_overlay_cflags,
-- 
2.39.1

